### PR TITLE
Bug fix: allow Basic Auth with custom auth header

### LINKED ISSUE #1043: Authorization header does not work for basic authentication
### Steps to reproduce:
use x-wso2-auth-header extension to define an authorization header
x-wso2-auth-header : "x-authorization"

use this auth header to send the request. it gets the below error without processing from any auth filter.
`"fault": {
        "code": 900901,
        "message": "Invalid Credentials",
        "description": "Invalid Credentials. Make sure you have given the correct access token"
    }`

### Environment details (with versions):
- OS:
- Client:
- Env (Docker/K8s):

---

### PR DESCRIPTION
**Pull request by VirajSalaka**
### Purpose
Bug fix: Allow basic auth to work with custom auth header. 
Add integration tests for basic auth scopes.  Related to #1103.
> At the moment, the basic Auth security requirement is added along with the oauth2 security requirement with scopes. According to the current ballerina implementation (as mentioned in #1103), those scopes will be checked during the basic authentication as well. At the moment, there is no approach to add the scopes in basic auth scenario via basic Auth SecurityRequirement/SecurityScheme alone.

### Issues

Fixes #1043 

### Automation tests
 - Unit tests added: No
 - Integration tests added: Yes

### Tested environments

MacOS Mojave
---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    # Checks if the request can be authenticated with the Basic Auth header.
+    #
+    # + req - The `Request` instance.
+    # + return - Returns `true` if can be authenticated. Else, returns `false`.
+    public function canProcess(http:Request req) returns @tainted boolean {
+        string authHeader = runtime:getInvocationContext().attributes[AUTH_HEADER].toString().trim();
+        //if '/apikey' resource is called, the handler is hit without the preAuthnFilter.
+        if (stringutils:equalsIgnoreCase("", authHeader)) {
+            authHeader = "authorization";
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal]: Use the constant for the header

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    public function __init(BasicAuthProvider basicAuthProvider) {
+        self.basicAuthProvider = basicAuthProvider;
+    }
+
+    # Checks if the request can be authenticated with the Basic Auth header.
+    #
+    # + req - The `Request` instance.
+    # + return - Returns `true` if can be authenticated. Else, returns `false`.
+    public function canProcess(http:Request req) returns @tainted boolean {
+        string authHeader = runtime:getInvocationContext().attributes[AUTH_HEADER].toString().trim();
```
Comment (Rajith90) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal]: There is a method to get authorization header from the service annotation.  Is that method setting the value to the context

Comment (VirajSalaka) [reply components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal]: In the pre authentication filter, it is added to the invocation context. Besides that, it is required to pass the filter context to call that specific method.

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal (hunk 1 above)
Comment (VirajSalaka) [reply components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal]: fixed via https://github.com/wso2/product-microgateway/pull/1132/commits/5f2b131876a0349201092db9985824582f120fba

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import ballerina/runtime;
+import ballerina/stringutils;
+
+# Representation of the basic auth handler
+#
+# + basicAuthProvider - The reference to the basic auth provider instance
+public type BasicAuthHandler object {
+
+    *http:InboundAuthHandler;
+
```
Comment (Shehanir) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal:28]: It is better to remove extra lines

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        string headerValue = req.getHeader(authHeader);
+        string credential = headerValue.substring(5, headerValue.length()).trim();
+        var authenticationResult = self.basicAuthProvider.authenticate(credential);
+        if (authenticationResult is boolean) {
+            return authenticationResult;
+        } else {
+            return prepareAuthenticationError("Failed to authenticate with basic auth handler.", authenticationResult);
+        }
+    }
+};
```
Comment (Shehanir) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal:76]: Shall we keep an extra line at EOF

Comment (VirajSalaka) [reply components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/basic_auth_handler.bal:76]: fix is already merged from some other PR.

**Pull request merged**